src/mpg123/mpg123.1
.\" @(#)mpg123.1 0.01 21-Apr-1997 OF; from mpg123 archive
.TH mpg123 1 "21 Apr 1997"
.SH NAME
mpg123 \- play audio MPEG 1.0/2.0 file (layers 1, 2 and 3)
.SH SYNOPSIS
.B mpg123
[
.B \-tscvqy01m24
]
[
.BI \-b " size"
]
[
.BI \-k " num"
]
[
.BI \-n " num"
]
[
.BI \-f " factor"
]
[
.BI \-r " rate"
]
[
.BI \-g " gain"
]
[
.BI \-a " dev"
]
[
.BR "\-o s" " | " "\-o h" " | " "\-o l"
]
[
.BI \-d " n"
]
[
.BI \-h " n"
]
[
.BI \-p " proxy"
]
[
.BI \-@ " file"
]
.IR file " ... | " URL " ... | "
.B \-
.SH DESCRIPTION
.B mpg123
reads one or more
.IR file\^ s
(or standard input if ``\-'' is specified) or
.IR URL\^ s
and plays them on the audio device (default) or
outputs them to stdout.
.IR file\^ / URL
is assumed to be an MPEG-1/2 audio bit stream.
.SH OPTIONS
.B mpg123
options may be either the traditional POSIX one letter options,
or the GNU style long options.  POSIX style options start with a
single ``\-'', while GNU long options start with ``\-\^\-''.
.TP
.BR \-t ", " \-\^\-test
Test mode.  The audio stream is decoded, but no output occurs.
.TP
.BR \-s ", " \-\^\-stdout
The decoded audio samples are written to standard output,
instead of playing them through the audio device.  This
option must be used if your audio hardware is not supported
by
.BR mpg123 .
The output format is raw (headerless) linear PCM audio data,
16 bit, stereo, host byte order.
.TP
.BR \-c ", " \-\^\-check
Check for filter range violations, and report them for each frame
if any occur.
.TP
.BR \-v ", " \-\^\-verbose
Increase the verbosity level.  For example, displays the frame
numbers during decoding.
.TP
.BR \-q ", " \-\^\-quiet
Quiet.  Suppress diagnostic messages.
.TP
.BR \-y ", " \-\^\-resync
Try to resync and continue decoding if an error occurs in
the input file.  Also try to recover from certain broken
headers.  Useful if you have a broken MPEG file, on which
.B mpg123
normally gives up saying `Illegal header'.  Be careful:
Broken locations in MPEG files might cause sharp, loud pops or
clicks, which might damage your speakers if played too loud.
.TP
.BR \-0 ", " \-\^\-single0 "; " \-1 ", " \-\^\-single1
Decode only channel 0 (left) or channel 1 (right),
respectively.  These options are available for
stereo MPEG streams only.
.TP
.BR \-m ", " \-\^\-singlemix
Mix both channels.  This option is available for
stereo MPEG layer-3 streams only.  It takes less
CPU time than full stereo decoding.
.TP
.BR \-2 ", " \-\^\-2to1 "; " \-4 ", " \-\^\-4to1
Performs a downsampling of ratio 2:1 (22 kHz) or 4:1 (11 kHz) 
on the output stream, respectively. Saves some CPU cycles, but 
at least the 4:1 ratio sounds ugly.
.TP
\fB\-b \fIsize\fR, \fB\-\^\-buffer \fIsize
Use an audio output buffer of
.I size
Kbytes.  This is useful to bypass short periods of heavy
system activity, which would normally cause the audio output 
to be interrupted.  
You should specify a buffer size of at least 1024 
(i.e. 1 Mb, which equals about 6 seconds of audio data) or more; 
less than about 300 does not make much sense.  The default is 0, 
which turns buffering off.
.TP
\fB\-k \fInum\fR, \fB\-\^\-skip \fInum
Skip first
.I num
frames.  By default the decoding starts at the first frame.
.TP
\fB\-n \fInum\fR, \fB\-\^\-frames \fInum
Decode only
.I num
frames.  By default the complete stream is decoded.
.TP
\fB\-f \fIfactor\fR, \fB\-\^\-scale \fIfactor
Change scale factor (default: 32768).
.TP
\fB\-r \fIrate\fR, \fB\-\^\-rate \fIrate
Set sample rate (default: automatic).  You don't want to
change this.
.TP
\fB\-g \fIgain\fR, \fB\-\^\-gain \fIgain
Set audio hardware output gain (default: don't change).
.TP
\fB\-a \fIdev\fR, \fB\-\^\-audiodevice \fIdev
Specify the audio device to use.  The default is
system-dependent (usually /dev/audio or /dev/dsp).
Use this option if you have multiple audio devices and
the default is not what you want.
.TP
.BR "\-o s" ", " \-\^\-speaker
Direct audio output to the speaker.
.TP
.BR "\-o h" ", " \-\^\-headphones
Direct audio output to the headphone connector.
.TP
.BR "\-o l" ", " \-\^\-lineout
Direct audio output to the line-out connector.
.TP
\fB\-d \fIn\fR, \fB\-\^\-doublespeed \fIn
Only play every
.IR n 'th
frame.  This will cause the MPEG stream
to be played
.I n
times faster, which can be used for special
effects.  Can also be combined with the
.B \-\^\-halfspeed
option to play 3 out of 4 frames etc.  Don't expect great
sound quality when using this option.
.TP
\fB\-h \fIn\fR, \fB\-\^\-halfspeed \fIn
Play each frame
.I n
times.  This will cause the MPEG stream
to be played at
.RI 1 / n 'th
speed (n times slower), which can be
used for special effects. Can also be combined with the
.B \-\^\-doublespeed
option to double every third frame or things like that.
Don't expect great sound quality when using this option.
.TP
\fB\-p \fIURL \fR| \fBnone\fR, \fB\-\^\-proxy \fIURL \fR| \fBnone
The specified
.I proxy
will be used for HTTP requests.  It
should be specified as full URL (``http://host.domain:port/''),
but the ``http://'' prefix, the port number and the trailing
slash are optional (the default port is 80).  Specifying
.B none
means not to use any proxy, and to retrieve files directly
from the respective servers.  See also the
``HTTP SUPPORT'' section.
.TP
\fB\-@ \fIfile\fR, \fB\-\^\-list \fIfile
Read filenames and/or URLs of MPEG audio streams from the specified
.I file
in addition to the ones specified on the command line (if any).
Note that
.I file
can be either an ordinary file, a dash ``\-'' to indicate that
a list of filenames/URLs is to be read from the standard input,
or an URL pointing to a an appropriate list file.  Note: only
one
.B \-@
option can be used (if more than one is specified, only the
last one will be recognized).
.SH OPERANDS
The following operands are supported:
.TP 8
.IR file (s)
The path name(s) of one or more input files.  They must be
valid MPEG-1/2 audio layer-1, -2 or -3 bit streams.
If a dash ``\-'' is specified, MPEG data will
be read from the standard input.  Furthermore, any name
starting with ``http://'' is recognized as
.I URL
(see next section).
.SH HTTP SUPPORT
In addition to reading MPEG audio streams from ordinary
files and from the standard input,
.B mpg123
supports retrieval of MPEG audio files via the HTTP protocol, 
which is used in the World Wide Web (WWW).  Such files are
specified using a so-called URL (universal resource
location), which starts with ``http://''.  When a file with
that prefix is encountered,
.B mpg123
attempts to open an HTTP connection to the server in order to
retrieve that file to decode and play it.
.P
It is often useful to retrieve files through a WWW cache or
so-called proxy.  To accomplish this,
.B mpg123
examines the environment for variables named
.BR MP3_HTTP_PROXY ", " http_proxy " and " HTTP_PROXY ,
in this order.  The value of the first one that is set will
be used as proxy specification.  To override this, you can
use the
.B \-p
command line option (see the ``OPTIONS'' section).  Specifying
.B "\-p none"
will enforce contacting the server directly without using
any proxy, even if one of the above environment variables
is set.
.P
Note that, in order to play MPEG audio files from a WWW
server, it is necessary that the connection to that server
is fast enough.  For example, a 128 kbit/s MPEG file
requires the network connection to be at least 128 kbit/s
(16 kbyte/s) plus protocol overhead.  If you suffer from
short network outages, you should try the
.B \-b
option (buffer) to bypass such outages.  If your network
connection is generally not fast enough to retrieve MPEG
audio files in realtime, you can first download the files
to your local harddisk (e.g. using
.BR lynx (1))
and then play them from there.
.SH INTERRUPT
You can abort
.B mpg123
at any time by pressing Ctrl-C.  If you are playing multiple
files, this will stop the current file and begin playing the
next one.  If you want to abort playing immediately instead
of skipping to the next file, press Ctrl-C twice in short
succession (within about one second).
.P
Note that the result of pressing Ctrl-C might not be audible
immediately, due to audio data buffering in the audio device.
This delay is system dependent, but it is usually not more
than one or two seconds.
.SH "SEE ALSO"
.BR lynx (1),
.BR sox (1),
.BR intro (1)
.SH NOTES
MPEG audio decoding requires a good deal of CPU performance,
especially layer-3.  To decode it in realtime, you should
have at least a Pentium, Alpha, SuperSparc or equivalent
processor.  You can also use the
.B -singlemix
option to decode mono only, which reduces the CPU load
somewhat for layer-3 streams.  See also the
.BR \-2 " and " \-4
options.
.P
If everything else fails, use the
.B \-s
option to decode to standard output, direct it into a file
and then use an appropriate utility to play that file.
You might have to use a tool such as
.BR sox (1)
to convert the output to an audio format suitable for
your audio player.
.P
Also note that
.B mpg123
always generates 16 bit stereo data (if one of the
.BR \-single *
options is used, two identical stereo channels are
generated).  If your hardware requires some other
format, for example 8 bit mono, you also have
to use a converter such as
.BR sox (1).
.P
If your system is generally fast enough to decode in 
realtime, but there are sometimes periods of heavy 
system load (such as cronjobs, users logging in remotely, 
starting of ``big'' programs etc.) causing the 
audio output to be interrupted, then you should use
the
.B \-b
option to use a buffer of at least 1000 Kbytes.
.SH BUGS
.TP
Known bugs and limitations:
.br
MPEG-2, Layer 1 and 2 not tested. May not work. (Layer 3 should work.)
.br
Free format streams are not supported.
.br
Layer-1 support is not heavily tested.
.br
No CRC error checking is performed.
.br
There is currently no support for audio hardware on
DEC Digital Unix, Ultrix and IBM AIX, therefore the
.B \-s
option has to be used on those platforms.
.SH AUTHORS
.TP
Main author:
.br
Michael Hipp <Michael.Hipp@student.uni-tuebingen.de>
.TP
Uses code (or at least ideas) from:
.br
MPEG Software Simulation Group (Base package)
.br
Philipp Knirsch <phil@mpik-tueb.mpg.de> (DCT36/manual unroll)
.br
Tobias Bading <bading@cs.tu-berlin.de> (subband synthesis)
.br
Jeff Tsay <ctsay@pasteur.eecs.berkeley.edu> (DCT36)
.br
Thomas Woerner (SGI Audio)
.br
Damien Clermonte <clermond@esiee.fr> (HP-UX audio fixes)
.br
Oliver Fromme <oliver.fromme@heim3.tu-clausthal.de>
.P
Internet references:
.br
http://www.sfs.nphil.uni-tuebingen.de/~hipp/mpg123.html
.br
http://www.heim3.tu-clausthal.de/~olli/mpg123/
.br
(includes information about the mpg123 mailing list)
.P
The latest version is also available from here:
.br
ftp.tu-clausthal.de:/pub/unix/audio/mpg123
.br
http://ftp.tu-clausthal.de/pub/unix/audio/mpg123


src/mpg123/mpg123.c
/* 
 * Mpeg Layer audio decoder (see version.h for version number)
 * ------------------------
 * copyright (c) 1995,1996,1997 by Michael Hipp, All rights reserved.
 * See also 'README' !
 *
 */

#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>

/* #define SET_PRIO */

#include "mpg123.h"
#include "getlopt.h"

#include "version.h"

static void usage(char *dummy);
static void print_title(void);

static long rates[3][3] = { 
 { 32000,44100,48000 } ,
 { 16000,22050,24000 } ,
 {  8000,11025,12000 } 
};

int supported_rates = 0;

int outmode = DECODE_AUDIO;

char *listname = NULL;
long outscale  = 32768;
int checkrange = FALSE;
int tryresync  = TRUE;
int quiet      = FALSE;
int verbose    = 0;
int doublespeed= 0;
int halfspeed  = 0;
int shuffle = 0;
int change_always = 1;
int force_8bit = 0;
int force_frequency = -1;
long numframes = -1;
long startFrame= 0;
int usebuffer  = 0;
int frontend_type = 0;
int remote     = 0;
int buffer_fd[2];
int buffer_pid;

static void catch_child(void)
{
  while (waitpid(-1, NULL, WNOHANG) > 0);
}

static int intflag = FALSE;
static int remflag = FALSE;

static void catch_interrupt(void)
{
  intflag = TRUE;
}

static char remote_buffer[1024];
static struct frame fr;
static struct audio_info_struct ai;
txfermem *buffermem;
#define FRAMEBUFUNIT (18 * 64 * 4)

void print_rheader(struct frame *fr);

static void catch_remote(void)
{
    remflag = TRUE;
    intflag = TRUE;
    if(usebuffer)
        kill(buffer_pid,SIGINT);
}


char *handle_remote(void)
{
	switch(frontend_type) {
		case FRONTEND_SAJBER:
#ifdef FRONTEND
			control_sajber(&fr);
#endif
			break;
		case FRONTEND_TK3PLAY:
#ifdef FRONTEND
			control_tk3play(&fr);
#endif
			break;
		default:
			fgets(remote_buffer,1024,stdin);
			remote_buffer[strlen(remote_buffer)-1]=0;
  
			switch(remote_buffer[0]) {
				case 'P':
					return remote_buffer+1;        
			}

			if(usebuffer)
				kill(buffer_pid,SIGINT);
			break;
	}

	return NULL;    
}

void init_output(void)
{
  static int init_done = FALSE;

  if (init_done)
    return;
  init_done = TRUE;
#ifndef OS2
  if (usebuffer) {
    unsigned int bufferbytes;
    sigset_t newsigset, oldsigset;

    if (usebuffer < 32)
      usebuffer = 32; /* minimum is 32 Kbytes! */
    bufferbytes = (usebuffer * 1024);
    bufferbytes -= bufferbytes % FRAMEBUFUNIT;
    xfermem_init (&buffermem, bufferbytes, sizeof(ai.rate));
    pcm_sample = (unsigned char *) buffermem->data;
    pcm_point = 0;
    sigemptyset (&newsigset);
    sigaddset (&newsigset, SIGUSR1);
    sigprocmask (SIG_BLOCK, &newsigset, &oldsigset);
    catchsignal (SIGCHLD, catch_child);
    switch ((buffer_pid = fork())) {
      case -1: /* error */
        perror("fork()");
        exit(1);
      case 0: /* child */
        xfermem_init_reader (buffermem);
        buffer_loop (&ai, &oldsigset);
        xfermem_done_reader (buffermem);
        xfermem_done (buffermem);
        _exit(0);
      default: /* parent */
        xfermem_init_writer (buffermem);
        outmode = DECODE_BUFFER;
    }
  }
  else {
#endif
    if (!(pcm_sample = (unsigned char *) malloc(audiobufsize * 2))) {
      perror ("malloc()");
      exit (1);
#ifndef OS2
    }
#endif
  }

  if(outmode==DECODE_AUDIO) {
    if(audio_open(&ai) < 0) {
      perror("audio");
      exit(1);
    }
    /* audio_set_rate (&ai);  should already be done in audio_open() [OF] */
  }
}

char *get_next_file (int argc, char *argv[])
{
    static FILE *listfile = NULL;
    static char line[1024];

#ifdef SHUFFLESUPPORT
	static int ord[2048];
	int temp, randomized,pos;
	static char initialized=0;
	time_t t;
#endif

	if (remote)
		return handle_remote();

    if (listname || listfile) {
        if (!listfile) {
            if (!*listname || !strcmp(listname, "-")) {
                listfile = stdin;
                listname = NULL;
            }
            else if (!strncmp(listname, "http://", 7)) 
                listfile = http_open(listname);
            else if (!(listfile = fopen(listname, "rb"))) {
                perror (listname);
                exit (1);
            }
            if (verbose)
                fprintf (stderr, "Using playlist from %s ...\n",
                        listname ? listname : "standard input");
        }
        do {
            if (fgets(line, 1023, listfile)) {
                line[strcspn(line, "\t\n\r")] = '\0';
                if (line[0]=='\0' || line[0]=='#')
                    continue;
                return (line);
            }
            else {
                if (*listname)
                   fclose (listfile);
                listname = NULL;
                listfile = NULL;
            }
        } while (listfile);
    }

#ifdef SHUFFLESUPPORT
	if(!initialized){
		for(pos=0;pos<=argc; pos++)   ord[pos]=pos;
		initialized=1;
	}
	if(shuffle){
		fprintf(stderr, "\nShuffle play - %u file(s) in loop.\n", argc-loptind);
		srandom(time(&t));
		for(pos=loptind;pos<argc;pos++){
			randomized=(random()%(argc-pos))+pos;
			temp=ord[pos];
			ord[pos]=ord[randomized];
			ord[randomized]=temp;
		}
		shuffle=0;
	}
	if (loptind < argc)
		return (argv[ord[loptind++]]);
	return (NULL);
#else
    if (loptind < argc)
    	return (argv[loptind++]);
    return (NULL);
#endif
}

void set_synth (char *arg)
{
    if (*arg == '2') {
        fr.down_sample = 1;
    }
    else {
        fr.down_sample = 2;
    }
}

#ifdef VARMODESUPPORT
void set_varmode (char *arg)
{
    audiobufsize = ((audiobufsize >> 1) + 63) & 0xffffc0;
}
#endif

void set_output (char *arg)
{
    switch (*arg) {
        case 'h': ai.output = AUDIO_OUT_HEADPHONES; break;
        case 's': ai.output = AUDIO_OUT_INTERNAL_SPEAKER; break;
        case 'l': ai.output = AUDIO_OUT_LINE_OUT; break;
        default:
            fprintf (stderr, "%s: Unknown argument \"%s\" to option \"%s\".\n",
                prgName, arg, loptarg);
            exit (1);
    }
}

void set_verbose (char *arg)
{
    verbose++;
}

topt opts[] = {
    {'k', "skip",        GLO_ARG | GLO_NUM,  0, &startFrame, 0},
    {'a', "audiodevice", GLO_ARG | GLO_CHAR, 0, &ai.device,  0},
    {'2', "2to1",        0,          set_synth, 0,           0},
    {'4', "4to1",        0,          set_synth, 0,           0},
    {'t', "test",        0,                  0, &outmode, DECODE_TEST},
    {'s', "stdout",      0,                  0, &outmode, DECODE_STDOUT},
    {'c', "check",       0,                  0, &checkrange, TRUE},
    {'v', "verbose",     0,        set_verbose, 0,           0},
    {'q', "quiet",       0,                  0, &quiet,      TRUE},
    {'y', "resync",      0,                  0, &tryresync,  FALSE},
    {'0', "single0",     0,                  0, &fr.single,  0},
    {0,   "left",        0,                  0, &fr.single,  0},
    {'1', "single1",     0,                  0, &fr.single,  1},
    {0,   "right",       0,                  0, &fr.single,  1},
    {'m', "singlemix",   0,                  0, &fr.single,  3},
    {0,   "mix",         0,                  0, &fr.single,  3},
    {'g', "gain",        GLO_ARG | GLO_NUM,  0, &ai.gain,    0},
    {'r', "rate",        GLO_ARG | GLO_NUM,  0, &force_frequency,  0},
    {0,   "8bit",        0,                  0, &force_8bit, 1},
    {0,   "headphones",  0,                  0, &ai.output, AUDIO_OUT_HEADPHONES},
    {0,   "speaker",     0,                  0, &ai.output, AUDIO_OUT_INTERNAL_SPEAKER},
    {0,   "lineout",     0,                  0, &ai.output, AUDIO_OUT_LINE_OUT},
    {'o', "output",      GLO_ARG | GLO_CHAR, set_output, 0,  0},
    {'f', "scale",       GLO_ARG | GLO_NUM,  0, &outscale,   0},
    {'n', "frames",      GLO_ARG | GLO_NUM,  0, &numframes,  0},
#ifdef VARMODESUPPORT
    {'v', "var",         0,        set_varmode, &varmode,    TRUE},
#endif
    {'b', "buffer",      GLO_ARG | GLO_NUM,  0, &usebuffer,  0},
	{'R', "remote",      0,                  0, &remote,     TRUE},
    {'d', "doublespeed", GLO_ARG | GLO_NUM,  0, &doublespeed,0},
    {'h', "halfspeed",   GLO_ARG | GLO_NUM,  0, &halfspeed,  0},
    {'p', "proxy",       GLO_ARG | GLO_CHAR, 0, &proxyurl,   0},
    {'@', "list",        GLO_ARG | GLO_CHAR, 0, &listname,   0},
#ifdef SHUFFLESUPPORT
	/* 'z' comes from the the german word 'zufall' (eng: random) */
    {'z', "shuffle",         0,        0, &shuffle,    1},
#endif
    {'?', "help",        0,              usage, 0,           0},
    {0, 0, 0, 0, 0, 0}
};

/*
 *   Change the playback sample rate.
 */
static void reset_audio_samplerate(void)
{
	if (usebuffer) {
		/* wait until the buffer is empty,
		 * then tell the buffer process to
		 * change the sample rate.   [OF]
		 */
		while (xfermem_get_usedspace(buffermem)	> 0)
			if (xfermem_block(XF_WRITER, buffermem) == XF_CMD_TERMINATE) {
				intflag = TRUE;
				break;
			}
		buffermem->freeindex = -1;
		buffermem->readindex = 0; /* I know what I'm doing! ;-) */
		buffermem->freeindex = 0;
		if (intflag)
			return;
		memcpy (buffermem->metadata, &ai.rate, sizeof(ai.rate));
		kill (buffer_pid, SIGUSR1);
	}
	else if (outmode == DECODE_AUDIO) {
		/* audio_reset_parameters(&ai); */
		/*   close and re-open in order to flush
		 *   the device's internal buffer before
		 *   changing the sample rate.   [OF]
		 */
		audio_close (&ai);
		if (audio_open(&ai) < 0) {
			perror("audio");
			exit(1);
		}
	}
}

/*
 * play a frame read read_frame();
 * (re)initialize audio if necessary.
 */
void play_frame(int init,struct frame *fr)
{
	int clip;

	if((fr->header_change && change_always) || init) {
		int reset_audio = 0;

		if(remote)
			print_rheader(fr);

		if (!quiet && init)
			if (verbose)
				print_header(fr);
			else
				print_header_compact(fr);

		if(force_frequency < 0) {
			if(ai.rate != freqs[fr->sampling_frequency]>>(fr->down_sample)) {
				ai.rate = freqs[fr->sampling_frequency]>>(fr->down_sample);
				reset_audio = 1;
			}
		}
		else if(ai.rate != force_frequency) {
			ai.rate = force_frequency;
			reset_audio = 1;
		}
		init_output();
		if(reset_audio) {
			reset_audio_samplerate();
			if (intflag)
				return;
		}
	}

	if (fr->error_protection) {
		getbits(16); /* crc */
	}

	clip = (fr->do_layer)(fr,outmode,&ai);

#ifndef OS2
	if (usebuffer) {
		if (!intflag) {
			buffermem->freeindex =
				(buffermem->freeindex + pcm_point) % buffermem->size;
			if (buffermem->wakeme[XF_READER])
				xfermem_putcmd(buffermem->fd[XF_WRITER], XF_CMD_WAKEUP);
		}
		pcm_sample = (unsigned char *) (buffermem->data + buffermem->freeindex);
		pcm_point = 0;
		while (xfermem_get_freespace(buffermem) < (FRAMEBUFUNIT << 1))
			if (xfermem_block(XF_WRITER, buffermem) == XF_CMD_TERMINATE) {
				intflag = TRUE;
				break;
			}
		if (intflag)
			return;
	}
#endif

	if(clip > 0 && checkrange)
		fprintf(stderr,"%d samples clipped\n", clip);
}


int main(int argc, char *argv[])
{
    int result;
    unsigned long frameNum = 0;
    char *fname;
    struct timeval start_time, now;
    unsigned long secdiff;
	int init;

#ifdef OS2
        _wildcard(&argc,&argv);
#endif

#ifdef SET_PRIO
	int mypid = getpid();
	setpriority(PRIO_PROCESS,mypid,-20);
#endif

	if(!strcmp("sajberplay",argv[0]))
		frontend_type = FRONTEND_SAJBER;
	if(!strcmp("mpg123m",argv[0]))
		frontend_type = FRONTEND_TK3PLAY;

    fr.single = -1; /* both channels */
    fr.synth = synth_1to1;
    fr.down_sample = 0;

    ai.format = AUDIO_FORMAT_SIGNED_16;
    ai.gain = ai.rate = ai.output = -1;
    ai.device = NULL;
    ai.channels = 2;

    (prgName = strrchr(argv[0], '/')) ? prgName++ : (prgName = argv[0]);

    while ((result = getlopt(argc, argv, opts)))
      switch (result) {
        case GLO_UNKNOWN:
          fprintf (stderr, "%s: Unknown option \"%s\".\n", prgName, loptarg);
          exit (1);
        case GLO_NOARG:
          fprintf (stderr, "%s: Missing argument for option \"%s\".\n",
              prgName, loptarg);
          exit (1);
      }
    if (loptind >= argc && !listname && !frontend_type)
      usage(NULL);

    if (remote){
        verbose = 0;        
        quiet = 1;
        catchsignal(SIGUSR1, catch_remote);
        fprintf(stderr,"@R MPG123\n");        
    }

    if (!quiet)
      print_title();


    {
        int fmts;
        int i,j;

        struct audio_info_struct ai;

        audio_info_struct_init(&ai);
        if (outmode == DECODE_AUDIO) {
        	audio_open(&ai);
        	fmts = audio_get_formats(&ai);
        }
        else
        	fmts = AUDIO_FORMAT_SIGNED_16;

        supported_rates = 0;
        for(i=0;i<3;i++) {
          for(j=0;j<3;j++) {
            ai.rate = rates[i][j];
            if (outmode == DECODE_AUDIO)
              audio_rate_best_match(&ai);
            /* allow about 2% difference */
            if( ((rates[i][j]*98) < (ai.rate*100)) &&
                ((rates[i][j]*102) > (ai.rate*100)) )
              supported_rates |= 1<<(i*3+j);
          } 
        }
        
        if (outmode == DECODE_AUDIO)
          audio_close(&ai);

        if(!force_8bit && !(fmts & AUDIO_FORMAT_SIGNED_16))
            force_8bit = 1;

        if(force_8bit && !(fmts & AUDIO_FORMAT_ULAW_8)) {
            fprintf(stderr,"No supported audio format found!\n");
            exit(1);
        }
    }

    if(force_8bit) {
#if 0
      ai.format = AUDIO_FORMAT_UNSIGNED_8;
      ai.format = AUDIO_FORMAT_SIGNED_8;
#endif
      ai.format = AUDIO_FORMAT_ULAW_8;
      make_conv16to8_table(ai.format);
      switch(fr.down_sample) {
        case 0:
          fr.synth = synth_1to1_8bit;
          fr.synth_mono = synth_1to1_8bit_mono;
          fr.block_size = 64;
          break;
        case 1:
          fr.synth = synth_2to1_8bit;
          fr.synth_mono = synth_2to1_8bit_mono;
          fr.block_size = 32;
          break;
        case 2:
          fr.synth = synth_4to1_8bit;
          fr.synth_mono = synth_4to1_8bit_mono;
          fr.block_size = 16;
          break;
      }
    }
    else {
      switch(fr.down_sample) {
        case 0:
          fr.synth = synth_1to1;
          fr.synth_mono = synth_1to1_mono;
          fr.block_size = 128;
          break;
        case 1:
          fr.synth = synth_2to1;
          fr.synth_mono = synth_2to1_mono;
          fr.block_size = 64;
          break;
        case 2:
          fr.synth = synth_4to1;
          fr.synth_mono = synth_4to1_mono;
          fr.block_size = 32;
          break;
      }
    }

    make_decode_tables(outscale);
    init_layer2(); /* inits also shared tables with layer1 */
    init_layer3(fr.down_sample);
    catchsignal (SIGINT, catch_interrupt);

	if(frontend_type) {
		handle_remote();
		exit(0);
	}

    while ((fname = get_next_file(argc, argv))) {
      char *dirname, *filename;

      if(!*fname || !strcmp(fname, "-"))
        fname = NULL;
      open_stream(fname,-1);
      
      if (!quiet) {
        if (split_dir_file(fname ? fname : "standard input",
                &dirname, &filename))
           fprintf(stderr, "\nDirectory: %s", dirname);
        fprintf(stderr, "\nPlaying MPEG stream from %s ...\n", filename);
      }

      gettimeofday (&start_time, NULL);
      read_frame_init();

      init = 1;
      for(frameNum=0;read_frame(&fr) && numframes && !intflag;frameNum++) {
			if(frameNum < startFrame || (doublespeed && (frameNum % doublespeed))) {
				if(fr.lay == 3)
					set_pointer(512);
				continue;
			}
			numframes--;
			play_frame(init,&fr);
			init = 0;
			if(verbose) {
				if (verbose > 1 || !(frameNum & 0xf))
					fprintf(stderr, "\r{%4lu} ",frameNum);
#ifndef OS2
			if (verbose > 1 && usebuffer)
				fprintf (stderr, "%7d ", xfermem_get_usedspace(buffermem));
#endif
			}

      }

      close_stream();
      if (!quiet) {
        /* This formula seems to work at least for
         * MPEG 1.0/2.0 layer 3 streams.
         */
        int sfd = freqs[fr.sampling_frequency] * (fr.lsf + 1);
        int secs = (frameNum * (fr.lay==1 ? 384 : 1152) + sfd / 2) / sfd;
        fprintf(stderr,"[%d:%02d] Decoding of %s finished.\n", secs / 60,
        	secs % 60, filename);
      }

	if(remote)
		fprintf(stderr,"@R MPG123\n");        
	if (remflag) {
		intflag = FALSE;
		remflag = FALSE;
	}

      if (intflag) {
        gettimeofday (&now, NULL);
        secdiff = (now.tv_sec - start_time.tv_sec) * 1000;
        if (now.tv_usec >= start_time.tv_usec)
          secdiff += (now.tv_usec - start_time.tv_usec) / 1000;
        else
          secdiff -= (start_time.tv_usec - now.tv_usec) / 1000;
        if (secdiff < 1000)
          break;
        intflag = FALSE;
      }
    }
#ifndef OS2
    if (usebuffer) {
      xfermem_done_writer (buffermem);
      waitpid (buffer_pid, NULL, 0);
      xfermem_done (buffermem);
    }
    else {
#endif
      audio_flush(outmode, &ai);
      free (pcm_sample);
#ifndef OS2
    }
#endif

    if(outmode==DECODE_AUDIO)
      audio_close(&ai);
    exit( 0 );
}

static void print_title(void)
{
    fprintf(stderr,"High Performance MPEG 1.0/2.0 Audio Player for Layer 1, 2 and 3.\n");
    fprintf(stderr,"Version %s (%s). Written and copyrights by Michael Hipp.\n", prgVersion, prgDate);
    fprintf(stderr,"Uses code from various people. See 'README' for more!\n");
    fprintf(stderr,"THIS SOFTWARE COMES WITH ABSOLUTELY NO WARRANTY! USE AT YOUR OWN RISK!\n");
}

static void usage(char *dummy)  /* print syntax & exit */
{
   print_title();
   fprintf(stderr,"\nusage: %s [option(s)] [file(s) | URL(s) | -]\n", prgName);
   fprintf(stderr,"supported options [defaults in brackets]:\n");
   fprintf(stderr,"   -v    increase verbosity level       -q    quiet (don't print title)\n");
   fprintf(stderr,"   -t    testmode (no output)           -s    write to stdout\n");
   fprintf(stderr,"   -k n  skip first n frames [0]        -n n  decode only n frames [all]\n");
   fprintf(stderr,"   -c    check range violations         -y    DISABLE resync on errors\n");
   fprintf(stderr,"   -b n  output buffer: n Kbytes [0]    -f n  change scalefactor [32768]\n");
   fprintf(stderr,"   -r n  override samplerate [auto]     -g n  set audio hardware output gain\n");
   fprintf(stderr,"   -os   output to built-in speaker     -oh   output to headphones\n");
   fprintf(stderr,"   -ol   output to line-out connector   -a d  set audio device\n");
   fprintf(stderr,"   -2    downsample 1:2 (22 kHz)        -4    downsample 1:4 (11 kHz)\n");
   fprintf(stderr,"   -d n  play every n'th frame only     -h n  play every frame n times\n");
   fprintf(stderr,"   -0    decode channel 0 (left) only   -1    decode channel 1 (right) only\n");
   fprintf(stderr,"   -m    mix both channels (mono)       -p p  use HTTP proxy p [$HTTP_PROXY]\n");
   fprintf(stderr,"   -@ f  read filenames/URLs from f     -z    shuffle play (with wildcards)\n");
   fprintf(stderr,"See the manpage %s(1) for more information.\n", prgName);
   exit(1);
}


